eclair-core/src/main/scala/fr/acinq/eclair/blockchain/electrum/ElectrumWallet.scala

```diff
@@ -557,77 +558,51 @@ object ElectrumWallet {
   case class TransactionConfidenceChanged(txid: ByteVector32, depth: Long, timestamp: Option[Long]) extends WalletEvent
   case class NewWalletReceiveAddress(address: String) extends WalletEvent
   case class WalletReady(confirmedBalance: Satoshi, unconfirmedBalance: Satoshi, height: Long, timestamp: Long) extends WalletEvent
-  // @formatter:on
 
-  /**
-   *
-   * @param key public key
-   * @return the address of the p2sh-of-p2wpkh script for this key
-   */
-  def segwitAddress(key: PublicKey, chainHash: ByteVector32): String = {
-    val script = Script.pay2wpkh(key)
-    val hash = Crypto.hash160(Script.write(script))
-    chainHash match {
-      case Block.RegtestGenesisBlock.hash | Block.TestnetGenesisBlock.hash => Base58Check.encode(Base58.Prefix.ScriptAddressTestnet, hash)
-      case Block.LivenetGenesisBlock.hash => Base58Check.encode(Base58.Prefix.ScriptAddress, hash)
-    }
-  }
-
-  def segwitAddress(key: ExtendedPrivateKey, chainHash: ByteVector32): String = segwitAddress(key.publicKey, chainHash)
-
-  def segwitAddress(key: PrivateKey, chainHash: ByteVector32): String = segwitAddress(key.publicKey, chainHash)
+  trait WalletType
+  case object P2SH_SEGWIT extends WalletType
+  case object NATIVE_SEGWIT extends WalletType
+  // @formatter:on
 
   /**
    *
-   * @param key public key
-   * @return a p2sh-of-p2wpkh script for this key
+   * @param scriptPubKey the public key script
+   * @return the hash of the public key script for this key, as used by Electrum's hash-based methods
    */
-  def computePublicKeyScript(key: PublicKey) = Script.pay2sh(Script.pay2wpkh(key))
+  def computeScriptHashFromScriptPubKey(scriptPubKey: Seq[ScriptElt]): ByteVector32 = Crypto.sha256(Script.write(scriptPubKey)).reverse
 
   /**
-   *
-   * @param key public key
-   * @return the hash of the public key script for this key, as used by Electrum's hash-based methods
-   */
-  def computeScriptHashFromPublicKey(key: PublicKey): ByteVector32 = Crypto.sha256(Script.write(computePublicKeyScript(key))).reverse
+    * We define a root path the first 3 elements in the BIP44 derivation scheme, which include
+    * the purpose, coin type (testnet/mainnet) and account number, example: m/purpose'/coin_type'/account'.
+    * We always use the first account.
+    * @return the root path
+    */
+  def rootPath(walletType: WalletType, chainHash: ByteVector32) = walletType match {
```

Comment:
It's quite hard to move it in the strategy since that _lives_ in `ElectrumWallet.Data`, this causes  a deadlock-like initialization error because [here](https://github.com/ACINQ/eclair/pull/1170/files#diff-2d0a62c2e3bc4b5904735f7004030ce7R115) we read the accountMaster/changeMaster before initializing the data. I'm moving the implementation into companion objects of the key store.

Reply:
Yes it's because some of the initial setup (deriving the first keys etc...) copies wallet logic instead of simply asking the wallet to do the actual work, I've a proposal to improve that (see comment below)

Reply:
Nice changes, i thought it was overkill to rework the initialization but it's a more elegant solution (and logically the accountPath goes in the key store). I will integrate that and the parametrized test as well